File: restaurant-management/java/src/test/java/com/restaurant/RestaurantServiceTest.java
package com.restaurant;

import com.restaurant.model.RestaurantTable;
import com.restaurant.model.Reservation;
import com.restaurant.repository.TableRepository;
import com.restaurant.repository.ReservationRepository;
import com.restaurant.service.RestaurantService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

@SpringBootTest
public class RestaurantServiceTest {

    @Autowired
    private TableRepository tableRepository;

    @Autowired
    private ReservationRepository reservationRepository;

    @Autowired
    private RestaurantService restaurantService;

    @BeforeEach
    public void setup() {
        reservationRepository.deleteAll();
        tableRepository.deleteAll();

        // Seedowanie stolików o różnych rozmiarach
        tableRepository.save(new RestaurantTable(2, "INDOOR"));
        tableRepository.save(new RestaurantTable(4, "INDOOR"));
        tableRepository.save(new RestaurantTable(4, "OUTDOOR"));
        tableRepository.save(new RestaurantTable(8, "INDOOR"));
    }

    @Test
    public void testOptimalTableBestFit() {
        // Szukamy stolika dla 3 osób o 18:00
        // Powinien wybrać stolik 4-osobowy, a nie 8-osobowy lub za mały 2-osobowy
        RestaurantTable optimal = restaurantService.findOptimalTable(3, "18:00");
        assertNotNull(optimal);
        assertEquals(4, optimal.getSeats());
    }

    @Test
    public void testReservedTableSkip() {
        // Pobieramy stoliki
        List<RestaurantTable> tables = tableRepository.findAll();
        RestaurantTable t4_1 = tables.stream().filter(t -> t.getSeats() == 4).toList().get(0);
        RestaurantTable t4_2 = tables.stream().filter(t -> t.getSeats() == 4).toList().get(1);

        // Zarezerwujmy oba stoliki 4-osobowe o 19:00
        reservationRepository.save(new Reservation(t4_1.getId(), "19:00", 3, "guest1"));
        reservationRepository.save(new Reservation(t4_2.getId(), "19:00", 4, "guest2"));

        // Szukamy dla 3 osób o 19:00. Skoro stoliki 4-osobowe są zajęte, optymalny musi być stolik 8-osobowy
        RestaurantTable optimal = restaurantService.findOptimalTable(3, "19:00");
        assertNotNull(optimal);
        assertEquals(8, optimal.getSeats());
    }

    @Test
    public void testNoMatchingTable() {
        // Szukamy stolika dla 10 osób
        RestaurantTable optimal = restaurantService.findOptimalTable(10, "20:00");
        assertNull(optimal);
    }

    @Test
    public void testBookTablePreventDuplicates() {
        List<RestaurantTable> tables = tableRepository.findAll();
        Long tId = tables.get(0).getId();

        // Pierwsza rezerwacja przechodzi
        String result1 = restaurantService.bookTable("guest1", tId, "20:00", 2);
        assertEquals("Reservation successful", result1);

        // Druga na ten sam stolik i czas zostaje odrzucona
        String result2 = restaurantService.bookTable("guest2", tId, "20:00", 2);
        assertEquals("Table already reserved", result2);
    }

    @Test
    public void testBookTablePreventUserDuplicates() {
        List<RestaurantTable> tables = tableRepository.findAll();
        Long tId1 = tables.get(0).getId();
        Long tId2 = tables.get(1).getId();

        // Pierwsza rezerwacja przechodzi
        String result1 = restaurantService.bookTable("guest1", tId1, "20:00", 2);
        assertEquals("Reservation successful", result1);

        // Druga rezerwacja tego samego użytkownika o tej samej godzinie zostaje odrzucona
        String result2 = restaurantService.bookTable("guest1", tId2, "20:00", 4);
        assertEquals("User already has a reservation at this time", result2);
    }
}

File: restaurant-management/nodejs/tests/restaurantService.test.js
const { findOptimalTable, bookTable } = require('../services/restaurantService');
const { sequelize, Table, Reservation } = require('../models');

beforeAll(async () => {
  await sequelize.sync({ force: true });
  
  // Seedowanie stolików o różnych rozmiarach
  await Table.bulkCreate([
    { id: 1, seats: 2, location: 'INDOOR' },
    { id: 2, seats: 4, location: 'INDOOR' },
    { id: 3, seats: 4, location: 'OUTDOOR' },
    { id: 4, seats: 8, location: 'INDOOR' },
  ]);
});

afterEach(async () => {
  await Reservation.destroy({ where: {} });
});

afterAll(async () => {
  await sequelize.close();
});

test('powinien dobrać najmniejszy wolny stolik pasujący do liczby gości (Best-Fit)', async () => {
  // Szukamy stolika dla 3 osób o 18:00
  // Powinien wybrać stolik 4-osobowy (id: 2 lub 3), a nie blokować 8-osobowego (id: 4) lub za małego 2-osobowego (id: 1)
  const optimal = await findOptimalTable(3, '18:00');
  expect(optimal).not.toBeNull();
  expect(optimal.seats).toBe(4);
  expect([2, 3]).toContain(optimal.id);
});

test('nie powinien dobrać zajętego stolika o podanej godzinie', async () => {
  // Zarezerwujmy stoliki 4-osobowe o 19:00
  await Reservation.create({ tableId: 2, time: '19:00', guestsCount: 3, userId: 'guest1' });
  await Reservation.create({ tableId: 3, time: '19:00', guestsCount: 4, userId: 'guest2' });

  // Szukamy dla 3 osób o 19:00. Skoro 4-osobowe są zajęte, optymalny musi być stolik 8-osobowy (id: 4)
  const optimal = await findOptimalTable(3, '19:00');
  expect(optimal).not.toBeNull();
  expect(optimal.id).toBe(4);
});

test('powinien zwrócić null jeśli nie ma stolika o wystarczającej pojemności', async () => {
  // Szukamy stolika dla 10 osób
  const optimal = await findOptimalTable(10, '20:00');
  expect(optimal).toBeNull();
});

test('powinien poprawnie zarezerwować stolik oraz odrzucić rezerwację na zajęty czas', async () => {
  const result1 = await bookTable('guest1', 1, '20:00', 2);
  expect(result1).toBe('Reservation successful');

  const result2 = await bookTable('guest2', 1, '20:00', 2);
  expect(result2).toBe('Table already reserved');
});
